src/facade/dragonfly_connection.h

```diff
@@ -444,8 +452,19 @@ class Connection : public util::Connection {
   // Per-thread queue backpressure structs.
   static thread_local QueueBackpressure tl_queue_backpressure_;
 
-  // a flag indicating whether the client has turned on client tracking.
-  bool tracking_enabled_ = false;
+  struct ClientTracking {
+    // a flag indicating whether the client has turned on client tracking.
+    bool tracking_enabled = false;
+    bool optin = false;
+    // remember if CLIENT CACHING TRUE was the last command
+    // true if prev command was CLIENT CACHING TRUE
+    bool prev_command = false;
```

Comment:
Yes! I will add a comment on the code and one here. So, we need to remember the last command executed because the chore with `OPTIN` is that if we call `CLIENT CACHING TRUE` then if the command that follows is a `read command` we track it.

Basically `OPTIN` is conditional caching on demand via `CLIENT CACHING TRUE` which acts as an `once flag` to potentially(since it might not be read only) track the next command.

I used `prev_command` and `last_command` because somehow I had to keep track the last executed command while keeping the change set to the code minimal (within `DispatchCommand`). 

The `prev_command` is the command executed in the previous call of `DispatchCommand`
The `last_command` is the command executed in the `current` call of `DispatchCommand`

If `last_command` is read only && `prev_command` is `CLIENT CACHING TRUE` then we need track the command (these are the semantics I described above).  At the end of the function `DisapatchCommand`, `prev_command` is updated to the 'last_command` and `last_command` is set to false (and only set to `true` when client caching true is called). 

p.s. The bonus is that this only requires to set the `last_command` to true on the call site of `CLIENT CACHING TRUE` avoiding us to explicitly set it to `false` for all other commands. We could do something similar by checking the `cid` and `args` in `DispatchCommand` but this introduces another layer of command parsing (name + args) on the dispatch layer which I am not happy about and for that reason I did not pursue

Reply:
Looking at this struct, maybe client tracking shouldn't be part of the connection code at all, like pubsub or monitoring? We can store information about it in connection state - like we do for other stuff. To track optin, we can introduce a seqnum for every command, and `ClientCaching()` will just do `conn_state.tracking.optin_seqnum = cntx->current_seqnum`

Reply:
maybe, in general I would like to refactor some parts of it. Shall we do this on separate PR to at least bake the basic functionality in ?

Reply:
Like you wish, but I think it can be done here 🤷🏻‍♂️

Reply:
@kostasrim I suspect that the state machine can be simplified, i.e. instead of last and prev fields, to have only
`next_command_tracking`. ClientCaching will set `next_command_tracking` and `DispatchCommand` will unconditionally exchange this var with `false` **before** executing the command. And I agree with @dranikpg  that this should be part of the context state like db_index or multi/exec state.

Reply:
@dranikpg @romange I refactored the code. However I still do think we need the two variables since there are two commands we need to track (the one executing at the moment via InvokeCmd and the one executed before it such that we can check if the last command was CACHING YES). I added a bunch of comments in the state maching, if you still think it';s doable with a single variable I am happy to reiterate :) It's a small change anyway :)